protocols/floodsub/src/protocol.rs

```diff
@@ -56,10 +63,28 @@ where
     type Error = FloodsubDecodeError;
     type Future = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;
 
-    fn upgrade_inbound(self, mut socket: TSocket, _: Self::Info) -> Self::Future {
+    fn upgrade_inbound(self, socket: TSocket, _: Self::Info) -> Self::Future {
         Box::pin(async move {
-            let packet = upgrade::read_length_prefixed(&mut socket, 2048).await?;
-            let rpc = rpc_proto::Rpc::decode(&packet[..]).map_err(DecodeError)?;
+            let mut framed = Framed::<TSocket, prost_codec::Codec<rpc_proto::Rpc>>::new(
+                socket,
+                prost_codec::Codec::new(MAX_MESSAGE_LEN_BYTES),
+            );
+            // let rpc = match framed.next().await {
+            //     Some(v) => v.unwrap(),
+            //     None => return DecodeError,
+            //     // _ => panic!(), // Better error here?
+            // };
+
+            // Better approach:
+            // let rpc = framed.next().await.unwrap_or_else(|| DecodeError);
+            // let rpc_a = framed.next().await.ok_or_else(|| DecodeError); //u nwrap_or_else(|e| (*e).clone());
+            // let rpc = rpc_a.unwrap();
+
+            // TODO: Understand what the ? is doing, and how to unwrap this without using a return (return in match causes errors in the map_err below.)
+            let rpc = framed
+                .next()
+                .await
+                .unwrap_or_else(|| FloodsubDecodeError::ReadError)?;
```

Comment:
You will have to construct an `io::Error` here to put into `ReadError`. "UnexpectedEOF" is likely the most appropriate.